lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +170,65 @@ def safe_remove_concurrent_index(table, options={})
     unsafe_remove_index(table, **options.merge(:algorithm => :concurrently))
   end
 
+  def safe_add_concurrent_partitioned_index(table, columns, name_suffix: nil, if_not_exists: false)
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    schema, table = _schema_and_table_for(table)
+
+    is_partitioned = select_value(<<~SQL)
+      SELECT EXISTS (
+        SELECT 1
+        FROM pg_partitioned_table, pg_class, pg_namespace
+        WHERE pg_class.oid = pg_partitioned_table.partrelid
+          AND pg_class.relnamespace = pg_namespace.oid
+          AND pg_class.relname = #{quote(table)}
+          AND pg_namespace.nspname = #{quote(schema)}
+      )
+    SQL
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{table} is not a partitioned table" unless is_partitioned
+
+    name_suffix = "#{Array.wrap(columns).join("_")}_idx" unless name_suffix.present?
+
+    child_tables = select_values(<<~SQL)
+      SELECT child.relname
+      FROM pg_inherits
+        JOIN pg_class parent ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child  ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace    ON parent.relnamespace = pg_namespace.oid
+      WHERE parent.relname = #{quote(table)}
+        AND pg_namespace.nspname = #{quote(schema)}
+    SQL
+
+    parent_index = "#{table}_#{name_suffix}"
+
+    quoted_columns = Array.wrap(columns).map { |c| quote_column_name(c) }.join(", ")
+
+    unsafe_execute(<<~SQL)
+      CREATE INDEX#{if_not_exists ? " IF NOT EXISTS" : ""} #{quote_column_name(parent_index)}
+      ON ONLY #{quote_schema_name(schema)}.#{quote_table_name(table)} (#{quoted_columns})
+    SQL
+
+    child_tables.each do |child_table|
+      child_index = "#{child_table}_#{name_suffix}"
+
+      next if if_not_exists && _index_exists?(schema, child_index)
```

Comment:
Can we use "normal" conditional style? I.e.:

```
unless if_not_exists && _index_exists?(schema, child_index)
  # do the thing
end
```

Also...do we want to use a real Postgres `IF NOT EXISTS`?

Reply:
I prefer to only use `unless` for single line conditionals. Otherwise it reads strangely in my head. But sure, I can try to rework this logic to avoid using `next`

I _think_ the `ALTER INDEX ... ATTACH PARTITION` call will fail if the index already exists and has been attached, which is why I chose to explicitly check for existence up front. I'll try to dig into that a bit more

Reply:
It turns out `ALTER INDEX ... ATTACH PARTITION` does _not_ fail if the index is already attached, so we _could_ use the native `IF NOT EXISTS`. It would just result in some extra queries. Do you think that's a reasonable tradeoff to simplify the code?

Reply:
Soooo I tried making this change but it appears there's a bug in rails 6.1 when generating the SQL (`CONCURRENTLY` should be before `IF NOT EXISTS`):

```
PG::SyntaxError:
  ERROR:  syntax error at or near "CONCURRENTLY"
  LINE 1: CREATE INDEX IF NOT EXISTS CONCURRENTLY "foos3_p2023w29_upda...
```